include/experimental/__p2630_bits/submdspan_extents.hpp

```diff
@@ -0,0 +1,332 @@
+template <size_t Counter, size_t... MapIdxs>
+struct inv_map_rank<Counter, index_sequence<MapIdxs...>> {
+  using type = index_sequence<MapIdxs...>;
+};
+
+#else
+// end of recursion specialization containing the final index_sequence
+template <size_t Counter, size_t... MapIdxs>
+MDSPAN_INLINE_FUNCTION
+auto inv_map_rank(integral_constant<size_t, Counter>, index_sequence<MapIdxs...>) {
+  return index_sequence<MapIdxs...>();
+};
+
+// specialization reducing rank by one (i.e., integral slice specifier)
+template<size_t Counter, class Slice, class... SliceSpecifiers, size_t... MapIdxs>
+MDSPAN_INLINE_FUNCTION
+auto inv_map_rank(integral_constant<size_t, Counter>, index_sequence<MapIdxs...>, Slice,
+                  SliceSpecifiers... slices) {
+  using next_idx_seq_t = conditional_t<is_convertible_v<Slice, size_t>,
+                                       index_sequence<MapIdxs...>,
+                                       index_sequence<MapIdxs..., Counter>>;
+
+  return inv_map_rank(integral_constant<size_t,Counter + 1>(), next_idx_seq_t(),
+                                     slices...);
+};
+#endif
+
+// Helper for identifying strided_index_range
+template <class T> struct is_strided_index_range : false_type {};
+
+template <class OffsetType, class ExtentType, class StrideType>
+struct is_strided_index_range<
+    strided_index_range<OffsetType, ExtentType, StrideType>> : true_type {};
+
+// first_of(slice): getting begin of slice specifier range
+MDSPAN_TEMPLATE_REQUIRES(
+  class Integral,
+  /* requires */(is_convertible_v<Integral, size_t>)
+)
+MDSPAN_INLINE_FUNCTION
+constexpr Integral first_of(const Integral &i) {
+  return i;
+}
+
+MDSPAN_INLINE_FUNCTION
+constexpr integral_constant<size_t, 0>
+first_of(const experimental::full_extent_t &) {
+  return integral_constant<size_t, 0>();
+}
+
+MDSPAN_TEMPLATE_REQUIRES(
+  class Slice,
+  /* requires */(is_convertible_v<Slice, tuple<size_t, size_t>>)
+)
+MDSPAN_INLINE_FUNCTION
+constexpr auto first_of(const Slice &i) {
+  return get<0>(i);
+}
+
+template <class OffsetType, class ExtentType, class StrideType>
+MDSPAN_INLINE_FUNCTION
+constexpr OffsetType
+first_of(const strided_index_range<OffsetType, ExtentType, StrideType> &r) {
+  return r.offset;
+}
+
+// last_of(slice): getting end of slice specifier range
+MDSPAN_TEMPLATE_REQUIRES(
+  size_t k, class Extents, class Integral,
+  /* requires */(is_convertible_v<Integral, size_t>)
+)
+MDSPAN_INLINE_FUNCTION
+constexpr Integral
+    last_of(integral_constant<size_t, k>, const Extents &, const Integral &i) {
+  return i;
+}
+
+MDSPAN_TEMPLATE_REQUIRES(
+  size_t k, class Extents, class Slice,
+  /* requires */(is_convertible_v<Slice, tuple<size_t, size_t>>)
+)
+MDSPAN_INLINE_FUNCTION
+constexpr auto last_of(integral_constant<size_t, k>, const Extents &,
+                       const Slice &i) {
+  return get<1>(i);
+}
+
+// Suppress spurious warning with NVCC about no return statement.
```

Comment:
I've seen this with NVCC + GCC 10.  The GCC intrinsic function `__builtin_unreachable();` has helped me get rid of the warnings in this case.  C++23 adds `std::unreachable`, which hopefully will accomplish the same goal.

Reply:
Does that GCC intrinsic also work with nvc++, clang, msvc as host compiler?

Reply:
It's a GCC-only thing, so when I used it, I defined it as a macro.

Reply:
GCC 10 isn't quite as smart as GCC 11 about `constexpr auto` functions that have exhaustive `if constexpr` branches in them.
```

Reply:
Thanks for figuring this out! : - )